In [2]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [8]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_df = pd.read_csv(
    Path("../Credit_Risk/Resources/lending_data.csv")
)

# Review the DataFrame
lending_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [11]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_df['loan_status']

# Separate the X variable, the features
X = lending_df.drop(columns=['loan_status'])

In [13]:
# Review the y variable Series
y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [14]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [17]:
# Check the balance of our target values
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [18]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [19]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
lr_model = logistic_regression_model.fit(X_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [43]:
# Make a prediction using the testing data

#Generate training predictions
#training_predictions = lr_model.predict(X_train)

#Generate testing predictions
testing_predictions = lr_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [64]:
# Print the balanced_accuracy score of the model
print("Accuracy Score:")
balanced_accuracy_score(y_test, testing_predictions)

Accuracy Score:


0.9520479254722232

In [62]:
# Generate a confusion matrix for the model
testing_matrix = confusion_matrix(y_test, testing_predictions)
print("Confusion Matrix:")
print(testing_matrix)

Confusion Matrix:
[[18663   102]
 [   56   563]]


In [65]:
# Print the classification report for the model
testing_report = classification_report(y_test, testing_predictions)
print(testing_report)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:**
- The logistic regression model performs exceptionally well for predicting healthy loans (label 0), with high precision and recall.
- For high-risk loans (label 1), the model has slightly lower precision but still performs well in terms of recall.
- The overall accuracy of the model is 99%, indicating its effectiveness in classifying both classes.

In conclusion, the logistic regression model demonstrates strong predictive performance for both healthy and high-risk loans, with particularly high accuracy for healthy loans.


---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [67]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros_module = RandomOverSampler(random_state = 1)

# Fit the original training data to the random_oversampler model
X_resampled, y_resampled = ros_module.fit_resample(X_train, y_train)

In [68]:
# Count the distinct values of the resampled labels data
resampled_label_counts = y_resampled.value_counts()

# Display the counts
print(resampled_label_counts)

0    56271
1    56271
Name: loan_status, dtype: int64


### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [69]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logreg_model = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
logreg_model.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
logreg_testing_predictions = logreg_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [70]:
# Print the balanced_accuracy score of the model 
print("Accuracy Score:")
balanced_accuracy_score(y_test, logreg_testing_predictions)

Accuracy Score:


0.9936781215845847

In [71]:
# Generate a confusion matrix for the model
testing_matrix = confusion_matrix(y_test, logreg_testing_predictions)
print("Confusion Matrix:")
print(testing_matrix)

Confusion Matrix:
[[18649   116]
 [    4   615]]


In [72]:
# Print the classification report for the model
testing_report = classification_report(y_test, logreg_testing_predictions)
print(testing_report)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 
- For healthy loans (label 0), the model achieves perfect precision, recall, and F1-score, indicating that it correctly identifies healthy loans with high accuracy.
- For high-risk loans (label 1), the model shows a slightly lower precision of 0.84, meaning that there are some false positives, but it has high recall (sensitivity) of 0.99, indicating that it effectively captures the majority of high-risk loans.

The overall accuracy of the model is 99%, and the macro and weighted averages of precision, recall, and F1-score are also high, suggesting that the model performs well across both classes.

In conclusion, the logistic regression model, trained on oversampled data, exhibits excellent performance in distinguishing between healthy and high-risk loans, achieving a high level of accuracy and effectively minimizing false negatives.